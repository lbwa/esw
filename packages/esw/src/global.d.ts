declare type TranspileModuleType = 'cjs' | 'esm' | 'iife'
declare type TranspileModeEnv = 'development' | 'production'
