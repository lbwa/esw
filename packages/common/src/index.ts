export * from './types'
export * from './load'
export * from './utils'
export * from './table'
export * from './serialize-size'
export * as stdout from './stdout'
export * from './esbuild-msg'
