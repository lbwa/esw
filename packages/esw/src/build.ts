import path from 'path'
import fs from 'fs'
import { build as esbuild, BuildOptions, BuildResult } from 'esbuild'
import { map, of, pipe, switchMap, tap } from 'rxjs'
import externalEsBuildPlugin from './plugins/external'

function normalizeBuildOptions(cwd: string) {
  return pipe(
    // resolve package.json from cwd
    map((raw: BuildOptions) => {
      const pkgJsonPath = path.resolve(cwd, './', 'package.json')
      if (!fs.existsSync(pkgJsonPath)) {
        throw new Error(`package.json is required in the ${pkgJsonPath}`)
      }
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const pkgJson = require(pkgJsonPath) as Record<string, unknown>
      const options: BuildOptions = {
        bundle: true,
        ...raw
      }
      return {
        options,
        pkgJson
      }
    }),
    // resolve outfile and outdir from package.json field
    tap(({ options, pkgJson }) => {
      const outPath = (pkgJson['main'] ?? pkgJson['module']) as string
      const outFile = path.basename(outPath).replace(path.extname(outPath), '')
      const outDir = path.basename(path.dirname(outFile))
      options.outfile ??= outFile
      options.outdir ??= outDir
    })
  )
}

function applyExternalPlugin() {
  return pipe(
    tap(
      ({
        options,
        pkgJson
      }: {
        options: BuildOptions
        pkgJson: Record<string, unknown>
      }) => {
        const peerDeps = pkgJson['peerDependencies'] as Record<string, string>
        const deps = pkgJson['dependencies'] as Record<string, string>
        const groups = ([] as string[]).concat(
          Object.keys(peerDeps),
          Object.keys(deps)
        )
        options.plugins = (options.plugins ?? []).concat(
          externalEsBuildPlugin(groups)
        )
      }
    )
  )
}

function invokeEsBuildBuild<V extends { options: BuildOptions }>() {
  return pipe(switchMap(({ options }: V) => esbuild(options)))
}

function build(
  options: BuildOptions = {},
  cwd: string = options.absWorkingDir || process.cwd()
) {
  const build$ = of(options).pipe(
    normalizeBuildOptions(cwd),
    applyExternalPlugin(),
    invokeEsBuildBuild()
  )
  return new Promise<BuildResult>((resolve, reject) => {
    build$.subscribe({
      next: resolve,
      error: reject
    })
  })
}

export default build
