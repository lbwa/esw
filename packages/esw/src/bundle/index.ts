export * from './service'
export { createInference } from './options'
