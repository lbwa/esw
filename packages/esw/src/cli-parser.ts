import path from 'path'
import arg from 'arg'
import {
  EMPTY,
  from,
  of,
  pipe,
  throwError,
  zip,
  tap,
  map,
  concatMap,
  catchError
} from 'rxjs'
import { printAndExit } from './shared/log'
import { ProcessCode as Code } from './shared/constants'

export type CommandRunner = (argv?: string[]) => void
type AvailableArgs = typeof availableArgs
type Commands = typeof COMMANDS

// eslint-disable-next-line
const pkgJson = require(path.resolve(__dirname, '../..', 'package.json'))
const DEFAULT_COMMAND_NAME = 'build'
const COMMANDS: Record<string, () => Promise<CommandRunner>> = {
  build: () => import('./cli/build').then(({ default: run }) => run)
}
const availableArgs = {
  '--version': Boolean,
  '--help': Boolean,

  // alias
  '-v': '--version',
  '-h': '--help'
}

function printVersion<V extends { args: arg.Result<AvailableArgs> }>(
  pkg: Record<string, unknown>
) {
  return pipe(
    tap(({ args }: V) => {
      if (args['--version']) {
        printAndExit(`v${pkg['version'] as string}`, Code.OK)
      }
    })
  )
}

function printHelpIntoTerminal(commands: Commands) {
  const names = Object.keys(commands)
  printAndExit(
    `
    Usage
      $ esw <command>

      Available commands
        ${names.join(', ')}

      Options
        --version, -v Show version number
        --help, -h Display help messages

      For more information run a command with the --help flag
        $ esw ${names[0] ?? ''} --help
  `,
    Code.OK
  )
}

function printHelp<
  V extends { isValidCommand: boolean; args: arg.Result<AvailableArgs> }
>(commands: Commands) {
  return pipe(
    tap(({ isValidCommand, args: { '--help': help } }: V) => {
      if (!isValidCommand && help) {
        printHelpIntoTerminal(commands)
      }
    })
  )
}

function setEnv<V extends { commandName: string }>() {
  return pipe(
    tap(({ commandName }: V) => {
      const normalizedEnv =
        commandName === 'build' ? 'production' : 'development'
      process.env['NODE_ENV'] ||= normalizedEnv

      process.on('SIGTERM', () => process.exit(0))
      process.on('SIGINT', () => process.exit(0))
    })
  )
}

function normalizeArgs<
  V extends { isValidCommand: boolean; args: arg.Result<AvailableArgs> }
>(defaultCommandName: string) {
  return pipe(
    map(({ isValidCommand, args }: V) => {
      const commandName = isValidCommand ? args._[0] : defaultCommandName
      const forwardArgs = isValidCommand ? args._.slice(1) : args._
      if (args['--help']) {
        forwardArgs.push('--help')
      }
      return {
        forwardArgs,
        commandName
      }
    })
  )
}

function invokeCommand<
  V extends { commandName: string; forwardArgs: string[] }
>() {
  return pipe(
    concatMap(({ commandName, forwardArgs }: V) =>
      zip(from(COMMANDS[commandName]()), of(forwardArgs))
    ),
    map(([run, args]) => run(args))
  )
}

export default of(process.argv.slice(2)).pipe(
  map(argv =>
    arg(availableArgs, {
      //https://github.com/vercel/arg/blob/5.0.0/index.js#L13
      argv,
      permissive: true
    })
  ),
  catchError((err: Error & { code: string }) => {
    if (err.code === 'ARG_UNKNOWN_OPTION') {
      printHelpIntoTerminal(COMMANDS)
      return EMPTY
    }
    return throwError(() => err)
  }),
  map(args => ({ isValidCommand: !!COMMANDS[args._[0]], args })),
  printVersion(pkgJson),
  printHelp(COMMANDS),
  normalizeArgs(DEFAULT_COMMAND_NAME),
  setEnv(),
  invokeCommand()
)
