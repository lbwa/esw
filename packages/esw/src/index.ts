export { default as build } from './build/node'
export { default as watch } from './watch/node'
