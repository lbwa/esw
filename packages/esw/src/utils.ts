export function printAndExit(msg: string, code = 1) {
  // eslint-disable-next-line no-console
  console[code === 0 ? 'log' : 'error'](msg)
  process.exit(code)
}
