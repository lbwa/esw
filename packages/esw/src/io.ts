import { promises as fs } from 'fs'
import path from 'path'
import { BuildOptions } from 'esbuild'
import { isDef } from '@eswjs/common'

export async function findAllStaleDir(options: BuildOptions[]) {
  const dirs = [] as string[]
  await Promise.all(
    options.map(async ({ outdir, absWorkingDir = process.cwd() }) => {
      if (
        !outdir ||
        path.isAbsolute(outdir) ||
        path.resolve(absWorkingDir, outdir) === path.resolve(absWorkingDir)
      )
        return

      const stat = await fs.lstat(outdir).catch(() => null)
      if (isDef(stat) && stat.isDirectory() && !dirs.includes(outdir)) {
        dirs.push(outdir)
      }
    })
  )
  return dirs
}

export function rmDirs(dirs: string[]) {
  return Promise.all(
    dirs.map(dir => fs.rm(dir, { recursive: true, force: true }))
  )
}

export async function isFileExists(filepath: string) {
  try {
    const stat = await fs.stat(filepath)
    return stat.isFile()
  } catch (error) {
    return false
  }
}
