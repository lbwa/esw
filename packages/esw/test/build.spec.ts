import path from 'path'
import fs from 'fs'
import { exec, ExecException } from 'child_process'
import { BuildOptions, BuildResult } from 'esbuild'
import { build } from '../src'

function getTestName() {
  return expect.getState().currentTestName.replace(/\s/g, '-')
}

function resolveFixture(name: string) {
  return path.resolve(__dirname, `./fixture/${name}`)
}

function clearCacheDir(dir: string) {
  return fs.promises.rm(dir, { force: true, recursive: true })
}

function formatBuildResult(raw: PromiseSettledResult<BuildResult>[]) {
  return raw.reduce((results, result) => {
    if (result.status === 'fulfilled') {
      results.push(result.value)
    }
    return results
  }, [] as BuildResult[])
}

function hasErrors(results: BuildResult[]) {
  return results.every(result => result.errors.length > 1)
}

function hasWarnings(results: BuildResult[]) {
  return results.every(result => result.warnings.length > 1)
}

async function runFixtureCase(
  fixtureName: string,
  cacheDir: string,
  options: BuildOptions,
  singleEntry = 'index.ts',
  outFile = 'index.js'
) {
  await clearCacheDir(resolveFixture(`${fixtureName}/${cacheDir}`))
  const results = await build({
    absWorkingDir: resolveFixture(fixtureName),
    logLevel: 'debug',
    format: 'esm',
    entryPoints: [singleEntry],
    outdir: cacheDir,
    bundle: false,
    ...options
  })
  const buildResults = formatBuildResult(results)
  expect(hasErrors(buildResults)).toBeFalsy()
  expect(hasWarnings(buildResults)).toBeFalsy()
  expect(buildResults).toMatchSnapshot(getTestName())

  expect(
    fs.existsSync(resolveFixture(`${fixtureName}/${cacheDir}/${outFile}`))
  ).toBeTruthy()

  return fs.promises.readFile(
    resolveFixture(`${fixtureName}/${cacheDir}/${outFile}`),
    { encoding: 'utf8' }
  )
}

beforeAll(async () => {
  await new Promise<ExecException | null>(r => {
    exec('yarn', { cwd: resolveFixture('typescript') }, r)
  })
})

describe('node api - build', () => {
  it('should work with package.json and dependencies', async () => {
    const cacheDir = 'dist/build'
    const outDir = resolveFixture(`typescript/${cacheDir}`)
    await clearCacheDir(outDir)
    const results = await build({
      absWorkingDir: resolveFixture('typescript'),
      logLevel: 'debug',
      outdir: cacheDir
    })
    const buildResults = formatBuildResult(results)
    expect(buildResults).toHaveLength(2)
    expect(hasErrors(buildResults)).toBeFalsy()
    expect(hasWarnings(buildResults)).toBeFalsy()
    expect(buildResults).toMatchSnapshot(getTestName())
    ;['', '.esm'].map(type =>
      expect(
        fs.existsSync(resolveFixture(`typescript/${cacheDir}/index${type}.js`))
      ).toBeTruthy()
    )

    const output = await fs.promises.readFile(
      resolveFixture(`typescript/${cacheDir}/index.esm.js`),
      { encoding: 'utf8' }
    )
    expect(output).toContain(`from "react"`)
    expect(output).toContain(`from "rxjs"`)
    expect(output).toContain(`from "rxjs/operators"`)

    const cjsOutput = await fs.promises.readFile(
      resolveFixture(`typescript/${cacheDir}/index.js`),
      { encoding: 'utf8' }
    )
    expect(cjsOutput).toContain(`require("react")`)
    expect(cjsOutput).toContain(`require("rxjs")`)
    expect(cjsOutput).toContain(`require("rxjs/operators")`)
  })

  it('should work with main field and cjs syntax', async () => {
    const output = await runFixtureCase('only-main-field', 'dist/cjs', {
      format: 'cjs'
    })
    expect(output).toContain(`require("./src/fib")`)
  })

  it('should work with main field and esm syntax', async () => {
    const output = await runFixtureCase('only-main-field', 'dist/esm', {
      // should respect option.format
      format: 'esm'
    })
    expect(output).toContain(`from "./src/fib"`)
  })

  it('should work with module field and esm syntax', async () => {
    const output = await runFixtureCase(
      'only-module-field',
      'dist',
      {},
      'index.ts',
      'index.esm.js'
    )
    expect(output).toContain(`from "./src/fib"`)
  })

  it("shouldn't override the esm inference from module field", async () => {
    const output = await runFixtureCase(
      'only-module-field',
      'dist',
      { format: 'cjs' },
      'index.ts',
      'index.esm.js'
    )
    expect(output).toContain(`from "./src/fib"`)
  })

  it('should override the cjs inference from main field', async () => {
    const output = await runFixtureCase(
      'only-module-field',
      'dist',
      { format: 'esm' },
      'index.ts',
      'index.esm.js'
    )
    expect(output).toContain(`from "./src/fib"`)
  })

  it("shouldn't work without package.json file", async () => {
    let err: Error | null = null
    await build({
      absWorkingDir: resolveFixture('required-package.json'),
      logLevel: 'debug',
      format: 'esm',
      entryPoints: ['index.ts'],
      outdir: 'dist',
      bundle: false
    }).catch((error: Error) => {
      err = error
    })
    expect(err).not.toBeNull()
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    expect(err!.message).toContain(`package.json file doesn't exists`)
  })

  it('should work with no options', async () => {
    const fixtureName = 'no-options'
    const cacheDir = 'dist/internal'
    await clearCacheDir(resolveFixture(`${fixtureName}/${cacheDir}`))
    const results = await build({
      absWorkingDir: resolveFixture(fixtureName),
      logLevel: 'debug'
    })
    const buildResults = formatBuildResult(results)
    expect(hasErrors(buildResults)).toBeFalsy()
    expect(hasWarnings(buildResults)).toBeFalsy()
    expect(buildResults).toMatchSnapshot(getTestName())

    const outFiles = ['index.js', 'index.esm.js'].map(filename =>
      fs.existsSync(resolveFixture(`${fixtureName}/${cacheDir}/${filename}`))
    )
    expect(outFiles.every(exists => exists)).toBeTruthy()

    const resolvedCjsOutFile = await fs.promises.readFile(
      resolveFixture(`${fixtureName}/${cacheDir}/index.js`),
      { encoding: 'utf8' }
    )
    expect(resolvedCjsOutFile).toContain(`__esModule"`)

    const resolvedEsmOutFile = await fs.promises.readFile(
      resolveFixture(`${fixtureName}/${cacheDir}/index.esm.js`),
      { encoding: 'utf8' }
    )
    expect(resolvedEsmOutFile).not.toContain(`__esModule`)
  })
})
