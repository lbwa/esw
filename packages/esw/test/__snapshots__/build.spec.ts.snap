// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`node api - build should override the cjs inference from main field: node-api---build-should-override-the-cjs-inference-from-main-field 1`] = `
Array [
  Object {
    "errors": Array [],
    "warnings": Array [],
  },
]
`;

exports[`node api - build should work with main field and cjs syntax: node-api---build-should-work-with-main-field-and-cjs-syntax 1`] = `
Array [
  Object {
    "errors": Array [],
    "warnings": Array [],
  },
]
`;

exports[`node api - build should work with main field and esm syntax: node-api---build-should-work-with-main-field-and-esm-syntax 1`] = `
Array [
  Object {
    "errors": Array [],
    "warnings": Array [],
  },
]
`;

exports[`node api - build should work with module field and esm syntax: node-api---build-should-work-with-module-field-and-esm-syntax 1`] = `
Array [
  Object {
    "errors": Array [],
    "warnings": Array [],
  },
]
`;

exports[`node api - build should work with no options: node-api---build-should-work-with-no-options 1`] = `
Array [
  Object {
    "errors": Array [],
    "warnings": Array [],
  },
  Object {
    "errors": Array [],
    "warnings": Array [],
  },
]
`;

exports[`node api - build should work with package.json and dependencies: node-api---build-should-work-with-package.json-and-dependencies 1`] = `
Array [
  Object {
    "errors": Array [],
    "warnings": Array [],
  },
  Object {
    "errors": Array [],
    "warnings": Array [],
  },
]
`;

exports[`node api - build shouldn't override the esm inference from module field: node-api---build-shouldn't-override-the-esm-inference-from-module-field 1`] = `
Array [
  Object {
    "errors": Array [],
    "warnings": Array [],
  },
]
`;
